File: lib/shim/speed_test.cpp
#include "speed_test.h"
#include "circbuf.h"
#include "spew.h"

#define DOT_SIZE 10
#define CIRC_NEW_ADD 3
#define CIRC_SUM_SIZE 7

static float a[DOT_SIZE];
static float b[DOT_SIZE];
static float accum;

static CircBuf cbuf(DOT_SIZE);

void cpp_speed_test_init() {
  for (int i = 0; i < DOT_SIZE;++i) {
    a[i] = (float) i;
    b[i] = (float) i;
  }
  accum = 0;
}

__attribute__((noinline)) float cpp_f32_dot() {
  float totes = 0.0;

  for (int i = 0; i < DOT_SIZE; ++i) {
    totes += (a[i] * b[i]);
  }

  accum = totes;
  return accum;
}

__attribute__((noinline)) float cpp_f32_circsum() {
  for (int i = 0; i < CIRC_NEW_ADD; ++i) {
    cbuf.push(a[i]);
  }
  int sum_offset = (DOT_SIZE - CIRC_SUM_SIZE) / 2;
  int sum_end = sum_offset + CIRC_SUM_SIZE;
  //assert(sum_offset >= 0 && sum_offset <= DOT_SIZE);
  //assert(sum_end >= 0 && sum_end <= DOT_SIZE);
  float totes = 0.0;
  for (int i = sum_offset; i < sum_end; ++i) {
    totes += cbuf.get(i);
  }
  accum = totes;
  return accum;
}

File: lib/shared/src/speed_test.rs
// use core::cell::OnceCell;

use crate::ds::circbuf::CircBuf;

#[allow(unused_imports)]
use crate::spew::*;

const DOT_SIZE: usize = 10;
#[allow(unused)]
const CIRC_NEW_ADD: usize = 3;
#[allow(unused)]
const CIRC_SUM_SIZE: usize = 7;

#[allow(non_upper_case_globals)]
static mut a: [f32; DOT_SIZE] = [0.0; DOT_SIZE];
#[allow(non_upper_case_globals)]
static mut b: [f32; DOT_SIZE] = [0.0; DOT_SIZE];
#[allow(non_upper_case_globals)]
static mut accum: f32 = 0.0;
#[allow(non_upper_case_globals)]
static mut cbuf: CircBuf::<f32> = CircBuf::<f32>::new_empty();

#[no_mangle]
pub fn rust_speed_test_init() {
    unsafe {
        for i in 0..DOT_SIZE {
            a[i] = i as f32;
            b[i] = i as f32;
        }
        accum = 0.0;
        cbuf = CircBuf::<f32>::new(DOT_SIZE, 0.0);
    }
}

#[no_mangle]
#[inline(never)]
pub fn rust_f32_dot() -> f32 {
    unsafe {
        let mut totes: f32 = 0.0;
        for i in 0..DOT_SIZE {
           totes += a[i] * b[i];
        }
        accum = totes;
        accum
    }
}

/*
#[no_mangle]
#[inline(never)]
pub fn rust_f32_circsum() -> f32 {
    unsafe {
        for i in 0..CIRC_NEW_ADD {
            cbuf.push(a[i]);
        }
        let sum_offset: usize = (DOT_SIZE - CIRC_SUM_SIZE) / 2;
        let sum_end: usize = sum_offset + CIRC_SUM_SIZE;
        //assert!(sum_offset <= DOT_SIZE);
        //assert!(sum_end <= DOT_SIZE);
        let mut totes: f32 = 0.0;
        for i in sum_offset..sum_end {
          totes += cbuf.get(i);
        }
        accum = totes;
        return accum;
    }
}
*/
